### PR TITLE
Update multi-host configuration

### PR DESCRIPTION
**Pull request by gabriel-v**
- set `[cluster] disable = all` as the default
- added test to run 3 servers on local network bridges
- filter health checks by the current node
- wait for duplicated system job health checks
- removed `source = /path/on/host` templates, as they can't work on multi-host
- fix a bug where the Docker container won't work outside of /opt/cluster

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -8,20 +8,15 @@ RUN set -e \
  && apt-get install  -qq -y --no-install-recommends \
     sudo curl unzip libcap2-bin qemu-kvm dnsutils iptables \
  && apt-get clean && rm -rf /var/lib/apt/lists/* \
- && mkdir -p /opt/cluster
+ && mkdir /src
 
-WORKDIR /opt/cluster
```
Comment (mgax) [inline Dockerfile]: What's wrong with `/opt/cluster`? If you really must change it, how about `/app`? `/src` feels like source code that's used to build the dist files.

File: Readme.md [hunk 2]
```diff
@@ -258,14 +258,22 @@ We also run some jobs as services:
...
 ```ini
 [cluster]
 disable = fabio,grafana
 ```
 
+For multi-host configurations one must disable running these jobs on all but one node, with:
+
+```ini
+cluster
+disable = all
```
Comment (mgax) [inline Readme.md]: We're entering double negation territory. How about replacing `disable` with `jobs` and you only enable the jobs you want?

File: ci/Vagrantfile [hunk 3]
```diff
@@ -11,7 +11,7 @@ Vagrant.configure("2") do |config|
   config.vm.synced_folder '.', '/vagrant', disabled: true
   config.vm.synced_folder "..", "/opt/cluster", type: "rsync",
       rsync__exclude: [".vagrant/", ".git/", "__pycache__/",
-                       "var/", "etc/"]
+                       "var/*", "etc/*", "bin/*"]
```
Comment (mgax) [inline ci/Vagrantfile]: So rsync does glob matching on full paths (relative to the sync root)? Maybe we can be even more strict, so we only ignore the `var` in the repo root?

File: ci/test-triple.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+echo "running three containers..."
+sudo mkdir /test
+sudo chown $(whoami): /test/
+for id in 1 2 4; do
+  cp -a . /test/$id
+  cp /test/$id/ci/configs/triple-$id.ini /test/$id/cluster.ini
+  /test/$id/examples/docker.sh --name test-$id
+done
+
+echo "wait until one of them wins"
```
Comment (mgax) [inline ci/test-triple.sh]: So the 3 containers are peers, with none of them being a special master? Then let's give them nearly-identical config files, without variation in nomad memory or `drain_on_stop`.

File: ci/test-triple.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  case "$count" in
+    "0") echo "no secret files" >&2; return 1 ;;
+    "1") echo "$onefile"; return 0 ;;
+    *) echo "too many vault-secrets.ini files!" >&2; exit 1
+  esac
+}
+until get_one_secret_file; do sleep 5; done
+winner=$(get_one_secret_file)
+
+echo "copy config over to the losers"
```
Comment (mgax) [inline ci/test-triple.sh]: Is there no better way to solve this? Maybe by having a shared volume where this file will be written, and using some kind of locking in autovault? I realise this is outside the scope of your PR, so maybe make an issue.

File: cluster.py [hunk 6]
```diff
@@ -48,10 +50,12 @@ class PATH:
     supervisord_sock = var / 'supervisor' / 'supervisor.sock'
 
 
+jinja_env = Environment(loader=FileSystemLoader(str(PATH.templates)))
```
Comment (mgax) [inline cluster.py]: Shocking!

File: cluster.py [hunk 7]
```diff
@@ -153,6 +170,9 @@ def validate(cls):
                 "cluster.ini: network.create_bridge must be unset on macOS"
             assert not cls.network_forward_ports, \
                 "cluster.ini: network.forward_ports must be unset on macOS"
+        if cls._disable and cls._disable not in (['all'], ['none']):
+            assert all(s in ALL_JOBS for s in cls._disable), \
+                'unidentified service in "cluster.disable" list'
```
Comment (mgax) [inline cluster.py]: A nicer way to solve this is to "parse" the _disable value, replace it with `ALL_JOBS` for `['all']`, `[]` for `['none']`, and then do a setwise substraction with `ALL_JOBS` to see if there's anything extra.

File: cluster.py [hunk 8]
```diff
@@ -166,8 +186,10 @@ def send(self, req):
                 res_body = json.load(res)
                 return res_body
 
-    def get(self, url):
-        return self.send(Request(f'{self.endpoint}{url}'))
+    def get(self, url, data=None):
```
Comment (mgax) [inline cluster.py]: Let's call it `params` here, because `data` means "body" for `put`.

File: cluster.py [hunk 9]
```diff
@@ -427,6 +448,13 @@ def supervisord(ctx, detach):
         sleep(1)
 
 
+def _supervisorctl(*args):
+    joined_args = " ".join(args)
+    conf = PATH.etc / "supervisord.conf"
+    subprocess.check_call(f'supervisorctl -c {conf} {joined_args}',
+                          shell=True)
```
Comment (mgax) [inline cluster.py]: No need for newline, it fits in 79 chars.

File: cluster.py [hunk 10]
```diff
@@ -427,6 +448,13 @@ def supervisord(ctx, detach):
         sleep(1)
 
 
+def _supervisorctl(*args):
+    joined_args = " ".join(args)
```
Comment (mgax) [inline cluster.py]: Yeah, this feels scary, why join them and do `shell=True` when you can pass them as a list directly?

File: cluster.py [hunk 11]
```diff
@@ -452,42 +480,66 @@ def supervisorctl(args):
...
+
+def get_failed_checks(health_checks, self_only, allow_duplicates):
+    """Generates a sequence of (service, check, status)
+    tuples for all failing checks after checking with Consul"""
+
+    def pick_worst(a, b):
+        for s in ['critical', 'warning', 'passing']:
+            if s in [a, b]:
+                return s
+        raise RuntimeError(f'unkown status: "{a}" and "{b}"')
```
Comment (mgax) [inline cluster.py]: Uppercase `Unknown` please :)

File: cluster.py [hunk 12]
```diff
@@ -563,37 +624,75 @@ def wait():
...
+        k: v for k, v in HEALTH_CHECKS.items()
+        if k in OPTIONS.get_jobs() and k in SYSTEM_JOBS
+    }, self_only=True, allow_duplicates=True)
+    wait_for_checks({
+        k: v for k, v in HEALTH_CHECKS.items()
+        if k in OPTIONS.get_jobs() and k not in SYSTEM_JOBS
+    })
+
+
+def restart_nomad_until_it_works():
```
Comment (mgax) [inline cluster.py]: :+1:

File: Readme.md (hunk 2 above)
Comment (gabriel-v) [reply Readme.md]: I agree, I was only hesitant because of the fact that we already rolled out code with that configuration name. I'll change it now

File: ci/Vagrantfile (hunk 3 above)
Comment (gabriel-v) [reply ci/Vagrantfile]: I don't want binaries and configs from my local repo flying into the one-time-use VM that the CI uses. It shouldn't matter, but it's cleaner this way.

File: ci/test-triple.sh (hunk 4 above)
Comment (gabriel-v) [reply ci/test-triple.sh]: That was the plan, the only difference should be `node_name` and `disable` (now `jobs`).

File: ci/test-triple.sh (hunk 5 above)
Comment (gabriel-v) [reply ci/test-triple.sh]: On a real multi-host setup this is a manual step since you need to access different machines. It should be in the readme. Here I'm just profiting from the fact that `autovault` behaves in this way, which simplifies testing.

If we only run 1 container in the beginning, Vault will never run because Consul won't have a leader (since its bootstrap count hasn't been met).

Comment (gabriel-v) [reply ci/test-triple.sh]: This will be addressed if/when we start providing Ansible / Puppet / Chef / * configurations to automate everything in here.

File: cluster.py (hunk 6 above)
Comment (gabriel-v) [reply cluster.py]: gotta have my `{% include`

File: Dockerfile (hunk 1 above)
Comment (gabriel-v) [reply Dockerfile]: This needs to be different from the mounts so the `bin/` folder remains unchanged (and with nothing mounted over it). Sure, `/app` it is

File: cluster.py (hunk 7 above)
Comment (gabriel-v) [reply cluster.py]: `_disable` gets set as a class field, so you can't really do anything in that class scope.
We'll refactor that later into a proper `__init__`

File: ci/Vagrantfile (hunk 3 above)
Comment (gabriel-v) [reply ci/Vagrantfile]: Now we're ignoring from the `.gitignore`.

File: cluster.py (hunk 7 above)
Comment (mgax) [reply cluster.py]: Ah, but you can treat class fields just like variables, overwrite them and everything.

File: ci/test-triple.sh (hunk 5 above)
Comment (mgax) [reply ci/test-triple.sh]: Yeah, we already have so much magic, maybe this is too much magic.

File: ci/Vagrantfile (hunk 3 above)
Comment (mgax) [reply ci/Vagrantfile]: Y'know, maybe we should renounce the `bin` folder entirely, have the dockerfile install consul+vault+nomad in /usr/local, and just tell people to always use docker.
